src/state/products.js

```diff
@@ -39,6 +39,39 @@ export const extractedProductShape = pt.shape({
   date: pt.string.isRequired,
 });
 
+/**
+ * Validate the contents of the given extracted product. This is used in lieu
+ * of proptypes since proptype checks do not work in a production build.
+ * @param {ExtractedProduct} extractedProduct
+ * @return {Boolean}
+ */
+export function isValidExtractedProduct(extractedProduct) {
+  if (typeof extractedProduct.title !== 'string') {
+    return false;
+  }
+
+  if (Number.isNaN(Date.parse(extractedProduct.date))) {
+    return false;
+  }
+
+  if (typeof extractedProduct.price !== 'number') {
+    return false;
+  }
+
+  for (const key of ['url', 'image']) {
+    try {
+      const url = new URL(extractedProduct[key]);
```

Comment:
Is converting the `url` value to an instance of the URL class how we are ensuring the value is "actually a URI", per the [Security Review](https://bugzilla.mozilla.org/show_bug.cgi?id=1497394#c12)? What is the risk here if we don't do this?

Reply:
> Is converting the `url` value to an instance of the URL class how we are ensuring the value is "actually a URI", per the [Security Review](https://bugzilla.mozilla.org/show_bug.cgi?id=1497394#c12)?

Yes, the URL constructor throws if the given string is not a valid URL.

> What is the risk here if we don't do this?

Besides needing to do this to correctly parse the URL and check that the host is http or https, we also want to ensure the value is a valid URL so that future operations can assume that it is a URL. 

[Defense in depth](https://en.wikipedia.org/wiki/Defense_in_depth_(computing)) requires that we add security measures at multiple levels, so even though we have a whitelist of allowed domains that we extract products from, we should assume the webpage and its contents are untrusted and may contain malicious HTML. If our browser action page treats `url` and `image` as URLs, and they turn out not to be valid URLs, there could be unintended behavior in the browser action page when it uses those properties in its HTML. Depending on the behavior, an attacker could use it as part of getting some JS run in a privileged extension context, as an example.

Validating the shape as being a valid URL basically cuts off any attacks that would rely on behavior stemming from invalid image URLs.